xds/src/main/java/io/grpc/xds/internal/sds/SdsProtocolNegotiators.java

```diff
@@ -338,29 +344,49 @@ public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
       checkNotNull(grpcHandler, "grpcHandler");
       this.grpcHandler = grpcHandler;
       this.downstreamTlsContext = downstreamTlsContext;
+      this.fallbackProtocolNegotiator = fallbackProtocolNegotiator;
     }
 
     @Override
     protected void handlerAdded0(final ChannelHandlerContext ctx) {
       final BufferReadsHandler bufferReads = new BufferReadsHandler();
       ctx.pipeline().addBefore(ctx.name(), null, bufferReads);
 
-      final SslContextProvider sslContextProvider =
-              TlsContextManagerImpl.getInstance()
-                      .findOrCreateServerSslContextProvider(downstreamTlsContext);
-
+      SslContextProvider sslContextProviderTemp = null;
+      try {
+        sslContextProviderTemp =
+            TlsContextManagerImpl.getInstance()
+                .findOrCreateServerSslContextProvider(downstreamTlsContext);
+      } catch (Exception e) {
+        if (fallbackProtocolNegotiator == null) {
+          ctx.fireExceptionCaught(new CertStoreException("No certificate source found!", e));
+          return;
+        }
+        logger.log(Level.INFO, "Using fallback for {0}", ctx.channel().localAddress());
+        // Delegate rest of handshake to fallback handler
+        ctx.executor()
```

Comment:
does this need to run in an executor? it seems like it should be fine without it.

Reply:
I had tried without the executor and I got the following exception. Probably because this was called inside `handlerAdded0`.


```
io.netty.channel.ChannelPipelineException: io.grpc.xds.internal.sds.SdsProtocolNegotiators$ServerSdsHandler.handlerAdded() has thrown an exception; removed.
	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:624)
	at io.netty.channel.DefaultChannelPipeline.replace(DefaultChannelPipeline.java:572)
	at io.netty.channel.DefaultChannelPipeline.replace(DefaultChannelPipeline.java:509)
	at io.grpc.xds.internal.sds.SdsProtocolNegotiators$HandlerPickerHandler.userEventTriggered(SdsProtocolNegotiators.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:332)
...
Caused by: java.lang.IllegalStateException: previous protocol negotiation event hasn't triggered
	at com.google.common.base.Preconditions.checkState(Preconditions.java:511)
	at io.grpc.netty.ProtocolNegotiators$ProtocolNegotiationHandler.fireProtocolNegotiationEvent(ProtocolNegotiators.java:765)
	at io.grpc.xds.internal.sds.SdsProtocolNegotiators$ServerSdsHandler.replaceHandler(SdsProtocolNegotiators.java:405)
	at io.grpc.xds.internal.sds.SdsProtocolNegotiators$ServerSdsHandler.handlerAdded0(SdsProtocolNegotiators.java:372)
	at io.grpc.netty.ProtocolNegotiators$ProtocolNegotiationHandler.handlerAdded(ProtocolNegotiators.java:726)
	... 36 more
```

Reply:
the problem here is you suppose to fire `ProtocolNegotiationEvent` when you received it from previous handler (unless it is the very first one). handlerAdded shouldn't just fire the PNE.

Reply:
Okay, fixed: by just calling replace to replace the current handler